base/array.jl

```diff
@@ -775,8 +775,13 @@ function collect(itr::Generator)
             return _array_for(et, itr.iter, isz)
         end
         v1, st = y
-        arr = _array_for(typeof(v1), itr.iter, isz, shape)
-        return collect_to_with_first!(arr, v1, itr, st)
+        dest = _array_for(typeof(v1), itr.iter, isz, shape)
+        # The typeassert gives inference a helping hand on the element type and dimensionality
+        # (work-around for #28382)
+        ElType = promote_typejoin_union(et)
+        ElType′ = ElType <: Type ? Type : ElType
```

Comment:
Seems like this should be part of the `et` computation at the top of the function (so that the total function here is more type-stable), no?

Reply:
Actually I did the same for `broadcast` (see [here](https://github.com/JuliaLang/julia/pull/30485/files#r243698858)) so that would make sense, even if that changes the return type in corner cases. I've just pushed a commit to do this.